In [1]:
import pandas as pd
pd.set_option('display.max_columns', None, 'display.max_rows', None)

In [2]:
fifa21 = pd.read_csv("Your folder path")
fifa21.shape

(18979, 77)

## Checking general information about the dataset such as: overall shape of data, types of the columns, and total NaN values in each columns.
## This helps in coming up with a general list of what to do in the cleaning process

In [3]:
fifa21.head(5)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   
4  €270K        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioni

In [4]:
fifa21.dtypes

ID                   int64
Name                object
LongName            object
photoUrl            object
playerUrl           object
Nationality         object
Age                  int64
↓OVA                 int64
POT                  int64
Club                object
Contract            object
Positions           object
Height              object
Weight              object
Preferred Foot      object
BOV                  int64
Best Position       object
Joined              object
Loan Date End       object
Value               object
Wage                object
Release Clause      object
Attacking            int64
Crossing             int64
Finishing            int64
Heading Accuracy     int64
Short Passing        int64
Volleys              int64
Skill                int64
Dribbling            int64
Curve                int64
FK Accuracy          int64
Long Passing         int64
Ball Control         int64
Movement             int64
Acceleration         int64
Sprint Speed         int64
A

In [5]:
fifa21.isna().sum()

ID                      0
Name                    0
LongName                0
photoUrl                0
playerUrl               0
Nationality             0
Age                     0
↓OVA                    0
POT                     0
Club                    0
Contract                0
Positions               0
Height                  0
Weight                  0
Preferred Foot          0
BOV                     0
Best Position           0
Joined                  0
Loan Date End       17966
Value                   0
Wage                    0
Release Clause          0
Attacking               0
Crossing                0
Finishing               0
Heading Accuracy        0
Short Passing           0
Volleys                 0
Skill                   0
Dribbling               0
Curve                   0
FK Accuracy             0
Long Passing            0
Ball Control            0
Movement                0
Acceleration            0
Sprint Speed            0
Agility                 0
Reactions   

In [6]:
fifa21.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

### Drop the Loan Data End column as it has too many NaN values 17966 from 18979 instances are NaNs

In [7]:
fifa21.drop(['Loan Date End'], axis= 1, inplace= True)
fifa21.shape

(18979, 76)

### Drop the useless columns: ID, Name, photoURL, playerURL

In [8]:
fifa21.drop(['ID', 'Name', 'photoUrl', 'playerUrl'], axis= 1, inplace= True)
fifa21.shape

(18979, 72)

### Converting the height column to ints and changing the feet to cm.
#### When we check for the unique values in the Height column we find that some cells are in the cm format and others are in the feet format. Plus being of object type we have to change either the cm to feet or the feet to cm and change the type of the data to numeric. In this case we will change all to the cm format as it is the standered unit for meacuring Height.

In [9]:
fifa21['Height'].unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

In [10]:
mask = fifa21['Height'].str.endswith('"')
df_feet = fifa21.loc[mask, 'Height']
fifa21['Height'] = fifa21['Height'].str.strip('cm')
fifa21.loc[mask, 'Height'] = df_feet.str.extract(r"(\d*)\'(\d*)").apply(pd.to_numeric, errors='coerce').mul([30.48, 2.54]).sum(axis=1).round().astype(int)

In [11]:
fifa21['Height'] = fifa21['Height'].astype(int)

In [12]:
fifa21['Height'].unique()

array([170, 187, 188, 181, 175, 184, 191, 178, 193, 185, 199, 173, 168,
       176, 177, 183, 180, 189, 179, 195, 172, 182, 186, 192, 165, 194,
       167, 196, 163, 190, 174, 169, 171, 197, 200, 166, 164, 198, 201,
       158, 162, 161, 160, 203, 157, 156, 202, 159, 206, 155])

### Converting the weight column to floats and changing the pounds to kg.
#### when we check for the unique values in the Weight column we find that some weights are in the kg format and other are in the lbs format. Plus they are all of object type as we can see in the cell bellow. It is also the same case as of the Height's column case. Here we will change all weights to kgs as it is the most common unit for meausring weights.

In [13]:
fifa21['Weight'].unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

In [15]:
mask = fifa21['Weight'].str.endswith('lbs')
fifa21['Weight'] = fifa21['Weight'].str.strip('kglbs')
df_lbs = fifa21[mask].astype(float)
fifa21.loc[mask, 'Weight'] = df_lbs['Weight'] * (1/2.2046)

In [16]:
fifa21['Weight'] = fifa21['Weight'].astype(int)

In [17]:
fifa21['Weight'].unique()

array([ 72,  83,  87,  70,  68,  80,  71,  91,  73,  85,  92,  69,  84,
        96,  81,  82,  75,  86,  89,  74,  76,  64,  78,  90,  66,  60,
        94,  79,  67,  65,  59,  61,  93,  88,  97,  77,  62,  63,  95,
       100,  58, 183, 179, 172, 196, 176, 185, 170, 203, 168, 161, 146,
       130, 190, 174, 148, 165, 159, 192, 181, 139, 154, 157, 163,  98,
       103,  99, 102,  56, 101,  57,  55, 104, 107, 110,  53,  50,  54,
        52])

### Now for the value, wage, and release clause columns they are of type object which needs to be changed to numeric.
#### As we see from the below cells that all 3 of those columns contain the M -> million, K -> thousand, and Euro symbol. We will strip the Euro symbol, change to numeric and multiply by million or thousand to get the desired value for each instance.

In [18]:
fifa21['Value'].unique()

array(['€103.5M', '€63M', '€120M', '€129M', '€132M', '€111M', '€120.5M',
       '€102M', '€185.5M', '€110M', '€113M', '€90.5M', '€82M', '€17.5M',
       '€83.5M', '€33.5M', '€114.5M', '€78M', '€103M', '€109M', '€92M',
       '€10M', '€76.5M', '€89.5M', '€87.5M', '€79.5M', '€124M', '€114M',
       '€95M', '€92.5M', '€105.5M', '€88.5M', '€85M', '€81.5M', '€26M',
       '€21M', '€56M', '€67.5M', '€53M', '€36.5M', '€51M', '€65.5M',
       '€46.5M', '€61.5M', '€72.5M', '€77.5M', '€43.5M', '€32.5M', '€36M',
       '€32M', '€54M', '€49.5M', '€57M', '€66.5M', '€74.5M', '€71.5M',
       '€121M', '€99M', '€67M', '€86.5M', '€93.5M', '€70M', '€62M',
       '€66M', '€58M', '€44M', '€81M', '€37M', '€14.5M', '€46M', '€47.5M',
       '€52.5M', '€54.5M', '€34.5M', '€57.5M', '€51.5M', '€44.5M', '€55M',
       '€48M', '€60.5M', '€63.5M', '€61M', '€29M', '€58.5M', '€55.5M',
       '€42M', '€40.5M', '€43M', '€45.5M', '€34M', '€26.5M', '€42.5M',
       '€35.5M', '€45M', '€41.5M', '€40M', '€11M', '€13.5M', '

In [19]:
fifa21['Wage'].unique()

array(['€560K', '€220K', '€125K', '€370K', '€270K', '€240K', '€250K',
       '€160K', '€260K', '€210K', '€310K', '€130K', '€350K', '€300K',
       '€190K', '€145K', '€195K', '€100K', '€140K', '€290K', '€82K',
       '€110K', '€230K', '€155K', '€200K', '€165K', '€95K', '€170K',
       '€105K', '€115K', '€150K', '€135K', '€55K', '€58K', '€81K', '€34K',
       '€120K', '€59K', '€90K', '€65K', '€56K', '€71K', '€18K', '€75K',
       '€47K', '€20K', '€84K', '€86K', '€74K', '€78K', '€27K', '€68K',
       '€85K', '€25K', '€46K', '€83K', '€54K', '€79K', '€175K', '€43K',
       '€49K', '€45K', '€38K', '€41K', '€39K', '€23K', '€51K', '€50K',
       '€87K', '€30K', '€14K', '€69K', '€31K', '€64K', '€53K', '€35K',
       '€21K', '€28K', '€17K', '€33K', '€70K', '€32K', '€89K', '€26K',
       '€40K', '€76K', '€72K', '€48K', '€36K', '€29K', '€60K', '€16K',
       '€37K', '€24K', '€52K', '€0', '€62K', '€73K', '€63K', '€19K',
       '€1K', '€66K', '€80K', '€12K', '€2K', '€42K', '€13K', '€900',
       '€5

In [20]:
fifa21['Release Clause'].unique()

array(['€138.4M', '€75.9M', '€159.4M', ..., '€59K', '€35K', '€64K'],
      dtype=object)

In [21]:
def money_columns_cleaning(df):
    df = df.str.strip('€')
    symbol_to_value = {'M':'*1000000', 'K': '*1000'}
    conversion = df.replace(symbol_to_value, regex= True).map(pd.eval)
    df = conversion.astype(int)
    return df

fifa21['Value'] = money_columns_cleaning(fifa21['Value'])
fifa21['Wage'] = money_columns_cleaning(fifa21['Wage'])
fifa21['Release Clause'] = money_columns_cleaning(fifa21['Release Clause'])

In [22]:
print(fifa21['Value'].unique()[:5])
print(fifa21['Wage'].unique()[:5])
print(fifa21['Release Clause'].unique()[:5])

[103500000  63000000 120000000 129000000 132000000]
[560000 220000 125000 370000 270000]
[138400000  75900000 159400000 161000000 166500000]


### Conversion of skill stars to number as ordinal data

In [23]:
cols = ['W/F', 'SM', 'IR']
print(fifa21['W/F'].unique())
print(fifa21['SM'].unique())
print(fifa21['IR'].unique())

['4 ★' '3 ★' '5 ★' '2 ★' '1 ★']
['4★' '5★' '1★' '2★' '3★']
['5 ★' '3 ★' '4 ★' '2 ★' '1 ★']


In [24]:
fifa21[cols] = fifa21[cols].apply(lambda x: x.str.replace('★', ''))
fifa21[cols] = fifa21[cols].apply(lambda x: x.str.replace(' ', ''))
print(fifa21['W/F'].unique())
print(fifa21['SM'].unique())
print(fifa21['IR'].unique())
fifa21[cols] = fifa21[cols].astype(int)    
fifa21[cols].dtypes

['4' '3' '5' '2' '1']
['4' '5' '1' '2' '3']
['5' '3' '4' '2' '1']


W/F    int32
SM     int32
IR     int32
dtype: object

### The Hits column had NaN values and numeric values but with K -> thousand.
#### So we removed the K and multiplied by 1000 and changed the column type to int

In [25]:
# mask = fifa21['Hits'].str.endswithwith('K')
fifa21['Hits'].fillna(0, inplace= True)
symbol_to_value = {'K': '*1000'}
fifa21['Hits'] = fifa21['Hits'].replace(symbol_to_value, regex= True).map(pd.eval).astype(int)

In [26]:
fifa21['Hits'].unique()

array([ 771,  562,  150,  207,  595,  248,  246,  120, 1600,  130,  321,
        189,  175,   96,  118,  216,  212,  154,  205,  202,  339,  408,
        103,  332,   86,  173,  161,  396, 1100,  433,  242,  206,  177,
       1500,  198,  459,  117,  119,  209,   84,  187,  165,  203,   65,
        336,  126,  313,  124,  145,  538,  182,  101,   45,  377,   99,
        194,  403,  414,  593,  374,  245, 3200,  266,  299,  309,  215,
        265,  211,  112,  337,   70,  159,  688,  116,   63,  144,  123,
         71,  224,  113,  168,   61,   89,  137,  278,   75,  148,  176,
        197,  264,  214,  247,  402,  440, 1700, 2300,  171,  320,  657,
         87,  259,  200,  255,  253,  196,   60,   97,   85,  169,  256,
        132,  239,  166,  121,  109,   32,   46,  122,   48,  527,  199,
        282,   51, 1900,  642,  155,  323,  288,  497,  509,   79,   49,
        270,  511,   80,  128,  115,  156,  204,  143,  140,  152,  220,
        134,  225,   94,   74,  135,  142,   50,   

### Some last modifications on the data
#### we will remove the new lines before the club names

In [27]:
fifa21['Club'] = fifa21['Club'].str.replace('\n\n\n\n','')
fifa21['Club'].head()

0           FC Barcelona
1               Juventus
2        Atlético Madrid
3        Manchester City
4    Paris Saint-Germain
Name: Club, dtype: object

#### Here we renamed the OVA column as it had an arrow before it and the LongName to Name 

In [28]:
fifa21.rename(columns={'↓OVA': 'OVA', 'LongName': 'Name'}, inplace=True)

In [29]:
fifa21.head()

Name Nationality  Age  OVA  POT  \
0                  Lionel Messi   Argentina   33   93   93   
1  C. Ronaldo dos Santos Aveiro    Portugal   35   92   92   
2                     Jan Oblak    Slovenia   27   91   93   
3               Kevin De Bruyne     Belgium   29   91   91   
4    Neymar da Silva Santos Jr.      Brazil   28   91   91   

                  Club     Contract   Positions  Height  Weight  \
0         FC Barcelona  2004 ~ 2021  RW, ST, CF     170      72   
1             Juventus  2018 ~ 2022      ST, LW     187      83   
2      Atlético Madrid  2014 ~ 2023          GK     188      87   
3      Manchester City  2015 ~ 2023     CAM, CM     181      70   
4  Paris Saint-Germain  2017 ~ 2022     LW, CAM     175      68   

  Preferred Foot  BOV Best Position        Joined      Value    Wage  \
0           Left   93            RW   Jul 1, 2004  103500000  560000   
1          Right   92            ST  Jul 10, 2018   63000000  220000   
2          Right   91            GK  Jul 16, 2014  120000000  125000   
3          Right   91           CAM  Aug 30, 2015  129000000  370000   
4          Right   91            LW   Aug 3, 2017  132000000  270000   

   Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0       138400000        429        85         95                70   
1        75900000        437        84         95                90   
2       159400000         95        13         11                15   
3       161000000        407        94         82                55   
4       166500000        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   
2        140          34             19           11      65         11   
3        408          76             66           88      94         84   
4        356          51             36           87      90         92   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   
2         68         57       27               12              18   
3         91        186       68               65              53   
4         93         94       35               30              29   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14

#### We changed the Joined column dtype from object to Datetime as it conveys the Date when the player joined the club according to the current contract

In [30]:
fifa21['Joined'] = pd.to_datetime(fifa21['Joined'], format='mixed')
fifa21['Joined'].head()

0   2004-07-01
1   2018-07-10
2   2014-07-16
3   2015-08-30
4   2017-08-03
Name: Joined, dtype: datetime64[ns]

### After we are done cleaning our data now we save it to another file so we can retrieve it easily if we need it in future analysis.

In [31]:
fifa21.to_csv('Your folder path')